src/framework/util/async_mutex.ts

```diff
@@ -0,0 +1,30 @@
+export class AsyncMutex {
+  private queue: Array<Promise<unknown>> = [];
+
+  // Run an async function with a lock on this mutex.
+  // Waits until the mutex is available, locks it, runs the function, then releases it.
+  async with<T>(fn: () => Promise<T>): Promise<T> {
```

Comment:
This looks correct, but is the queue really necessary? I think the AsyncMutex only needs to remember the last promise it generated, `await` it, and then replace the last promise.

Reply:
Good call, let me try to make that work.

Reply:
Much better, thanks!